src/NextGen/Gateways/Stripe/NextGenStripeGateway/Webhooks/Listeners/InvoicePaymentSucceeded.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace Give\NextGen\Gateways\Stripe\NextGenStripeGateway\Webhooks\Listeners;
+
+use Give\Donations\Models\Donation;
+use Give\Donations\Models\DonationNote;
+use Give\Donations\ValueObjects\DonationStatus;
+use Give\Donations\ValueObjects\DonationType;
+use Give\Framework\Exceptions\Primitives\Exception;
+use Give\Framework\Support\Facades\DateTime\Temporal;
+use Give\Framework\Support\ValueObjects\Money;
+use Give\NextGen\Gateways\Stripe\NextGenStripeGateway\NextGenStripeGateway;
+use Give\Subscriptions\Models\Subscription;
+use Give\Subscriptions\ValueObjects\SubscriptionStatus;
+use Stripe\Event;
+use Stripe\Invoice;
+
+/**
+ * @unreleased
+ */
+class InvoicePaymentSucceeded
+{
+    use StripeWebhookListenerRepository;
+
+    /**
+     * Processes invoice.payment_succeeded event.
+     *
+     * Occurs whenever an invoice payment attempt succeeds.
+     *
+     * @see https://stripe.com/docs/api/events/types#event_types-invoice.payment_succeeded
+     *
+     * @unreleased
+     *
+     * @return void
+     * @throws \Exception
+     */
+    public function __invoke(Event $event)
+    {
+        try {
+            $this->processEvent($event);
+        } catch (Exception $exception) {
+            $this->logWebhookError($event, $exception);
+        }
+
+        exit;
+    }
+
+    /**
+     * @unreleased
+     * @throws \Exception
+     */
+    public function processEvent(Event $event)
+    {
+        /* @var Invoice $invoice */
+        $invoice = $event->data->object;
+
+        $subscription = give()->subscriptions->queryByGatewaySubscriptionId($invoice->subscription)->get();
+
+        // only use this for next gen for now
+        if (!$subscription || !$this->shouldProcessSubscription($subscription)) {
+            return;
+        }
+
+        if ($initialDonation = give()->donations->getByGatewayTransactionId($invoice->payment_intent)) {
+            $this->handleInitialDonation($initialDonation);
+        } else {
+            if ($this->shouldCreateRenewal($subscription)) {
```

Comment:
This might be a good place for a decorator, so that we can pass the `$subscription` as a dependency of the decorator and then just call the `shouldCreateRenewal`, `shouldEndSubscription` checks without passing it every time.

Reply:
resolved 05e54e04 

note: I had to keep a local copy of the `cancelSubscription` method for mocking because it has side effects but moved everything else to the decorator.